// network-as-code/dist/esm/api/locationApi.js
/**
 * Copyright 2024 Nokia
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { errorHandler } from "../errors/index.js";
import fetch from "node-fetch";
class LocationVerifyAPI {
    constructor(baseUrl, rapidKey, rapidHost, agent) {
        this.baseUrl = baseUrl;
        this.headers = {
            "Content-Type": "application/json",
            "X-RapidAPI-Host": rapidHost,
            "X-RapidAPI-Key": rapidKey,
        };
        this.agent = agent;
    }
    verifyLocation(latitude_1, longitude_1, device_1, radius_1) {
        return __awaiter(this, arguments, void 0, function* (latitude, longitude, device, radius, maxAge = 60) {
            const body = {
                device: device.toJson(),
                area: {
                    areaType: "CIRCLE",
                    center: { latitude: latitude, longitude: longitude },
                    radius: radius,
                },
                maxAge,
            };
            const response = yield fetch(`${this.baseUrl}/verify`, {
                method: "POST",
                headers: this.headers,
                body: JSON.stringify(body),
                agent: this.agent,
            });
            errorHandler(response);
            const data = yield response.json();
            return data.verificationResult === "TRUE"
                ? true
                : data.verificationResult === "FALSE"
                    ? false
                    : data.verificationResult;
        });
    }
}
class LocationRetrievalAPI {
    constructor(baseUrl, rapidKey, rapidHost, agent) {
        this.baseUrl = baseUrl;
        this.headers = {
            "Content-Type": "application/json",
            "X-RapidAPI-Host": rapidHost,
            "X-RapidAPI-Key": rapidKey,
        };
        this.agent = agent;
    }
    getLocation(device_1) {
        return __awaiter(this, arguments, void 0, function* (device, maxAge = 60) {
            const body = { device: device.toJson(), maxAge };
            const response = yield fetch(`${this.baseUrl}/retrieve`, {
                method: "POST",
                headers: this.headers,
                body: JSON.stringify(body),
                agent: this.agent,
            });
            errorHandler(response);
            const jsonData = yield response.json();
            return {
                latitude: jsonData.area.center.latitude,
                longitude: jsonData.area.center.longitude,
                civicAddress: jsonData.civicAddress,
                radius: jsonData.area.radius,
            };
        });
    }
}
export { LocationRetrievalAPI, LocationVerifyAPI };
